File: backend/app/schemas/app_cloner.py
from datetime import datetime

from pydantic import BaseModel, Field


class AppCloneRequest(BaseModel):
    repo_url: str = Field(min_length=1, max_length=500)
    destination_parent: str = Field(min_length=1, max_length=500)
    directory_name: str | None = Field(default=None, max_length=120)
    run_npm_start: bool = True


class AppCloneStatus(BaseModel):
    running: bool
    status: str
    message: str
    repo_url: str | None = None
    destination_parent: str | None = None
    clone_path: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    git_exit_code: int | None = None
    npm_start_pid: int | None = None
    log_path: str
    last_lines: list[str] = []


class AppCloneDefaults(BaseModel):
    repo_url: str | None = None

File: frontend/src/api/appCloner.ts
import { buildApiUrl, parseJsonResponse } from "./client";

export type AppCloneStatus = {
  running: boolean;
  status: string;
  message: string;
  repo_url: string | null;
  destination_parent: string | null;
  clone_path: string | null;
  started_at: string | null;
  finished_at: string | null;
  git_exit_code: number | null;
  npm_start_pid: number | null;
  log_path: string;
  last_lines: string[];
};

export type AppCloneDefaults = {
  repo_url: string | null;
};

export type AppCloneRequest = {
  repo_url: string;
  destination_parent: string;
  directory_name?: string | null;
  run_npm_start: boolean;
};

export async function getAppCloneDefaults() {
  const response = await fetch(await buildApiUrl("/api/app-cloner/defaults"));
  return parseJsonResponse<AppCloneDefaults>(response);
}

export async function getAppCloneStatus() {
  const response = await fetch(await buildApiUrl("/api/app-cloner/status"));
  return parseJsonResponse<AppCloneStatus>(response);
}

export async function cloneApp(payload: AppCloneRequest) {
  const response = await fetch(await buildApiUrl("/api/app-cloner/clone"), {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  return parseJsonResponse<AppCloneStatus>(response);
}
